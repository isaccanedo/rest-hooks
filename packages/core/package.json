{
  "name": "@rest-hooks/core",
  "version": "3.4.0",
  "description": "Asynchronous data framework for React",
  "sideEffects": false,
  "main": "dist/index.js",
  "react-native": "legacy/index.js",
  "module": "legacy/index.js",
  "unpkg": "dist/index.umd.min.js",
  "types": "lib/index.d.ts",
  "typesVersions": {
    ">=4.0": {
      "": [
        "lib/index.d.ts"
      ],
      "*": [
        "lib/index.d.ts"
      ]
    },
    ">=3.4": {
      "": [
        "ts3.4/index.d.ts"
      ],
      "*": [
        "ts3.4/index.d.ts"
      ]
    }
  },
  "exports": {
    ".": {
      "module": "./lib/index.js",
      "import": "./node.mjs",
      "require": "./dist/index.js",
      "default": "./lib/index.js"
    },
    "./package.json": "./package.json"
  },
  "type": "module",
  "engines": {
    "node": "^12.17 || ^13.7 || >=14"
  },
  "files": [
    "src",
    "dist",
    "lib",
    "ts3.4",
    "node.mjs",
    "legacy",
    "LICENSE",
    "README.md",
    "./typescript.svg",
    "./rest_hooks_logo_and_text.svg"
  ],
  "scripts": {
    "build:lib": "NODE_ENV=production BROWSERSLIST_ENV='2020' babel --root-mode upward src --out-dir lib --source-maps inline --extensions '.ts,.tsx,.js,.jsx' --ignore '**/__tests__/**' --ignore '**/*.d.ts'",
    "build:legacy:lib": "NODE_ENV=production BROWSERSLIST_ENV='2018' babel --root-mode upward src --out-dir legacy --source-maps inline --extensions '.ts,.tsx,.js,.jsx' --ignore '**/__tests__/**' --ignore '**/*.d.ts'",
    "build:js:node": "BROWSERSLIST_ENV=node12 rollup -c",
    "build:js:browser": "BROWSERSLIST_ENV=legacy rollup -c",
    "build:bundle": "run-s build:js:\\* && echo '{\"type\":\"commonjs\"}' > dist/package.json",
    "build:clean": "rimraf lib ts3.4 legacy dist *.tsbuildinfo ./index.d.ts",
    "build:legacy-types": "yarn run downlevel-dts lib ts3.4 && copyfiles --up 1 ./src-legacy-types/**/*.d.ts ./ts3.4/",
    "build": "yarn run build:lib && yarn run build:legacy:lib && yarn run build:bundle",
    "dev": "yarn run build:lib -w",
    "prepare": "yarn run build:lib",
    "prepack": "yarn prepare",
    "prepublishOnly": "yarn run build:bundle && yarn run build:legacy:lib"
  },
  "keywords": [
    "react",
    "data",
    "cache",
    "flux",
    "suspense",
    "fetch",
    "hook",
    "networking",
    "async",
    "concurrent mode",
    "typescript",
    "async",
    "data fetching",
    "data cache",
    "api client",
    "api",
    "normalized cache",
    "swr",
    "GraphQL",
    "query",
    "react-native",
    "ios",
    "android",
    "web",
    "middleware",
    "declarative",
    "scalable"
  ],
  "author": "Nathaniel Tucker <me@ntucker.me> (https://github.com/ntucker)",
  "license": "Apache-2.0",
  "homepage": "https://resthooks.io",
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com:coinbase/rest-hooks.git",
    "directory": "packages/core"
  },
  "bugs": {
    "url": "https://github.com/coinbase/rest-hooks/issues"
  },
  "dependencies": {
    "@babel/runtime": "^7.13.0",
    "@rest-hooks/normalizr": "^9.2.0",
    "@rest-hooks/use-enhanced-reducer": "^1.2.0",
    "flux-standard-action": "^2.1.1"
  },
  "peerDependencies": {
    "@types/react": "^16.8.4 || ^17.0.0 || ^18.0.0-0",
    "react": "^16.8.4 || ^17.0.0 || ^18.0.0-0"
  },
  "peerDependenciesMeta": {
    "@types/react": {
      "optional": true
    }
  },
  "devDependencies": {
    "@babel/cli": "7.19.3",
    "@babel/core": "7.20.2",
    "@types/babel__core": "^7",
    "downlevel-dts": "^0.10.0",
    "npm-run-all": "^4.1.5",
    "rollup": "2.79.1",
    "rollup-plugin-babel": "^4.4.0",
    "rollup-plugin-commonjs": "^10.1.0",
    "rollup-plugin-dts": "^5.0.0",
    "rollup-plugin-filesize": "^9.1.2",
    "rollup-plugin-json": "^4.0.0",
    "rollup-plugin-node-resolve": "^5.2.0",
    "rollup-plugin-replace": "^2.2.0",
    "rollup-plugin-terser": "^7.0.2"
  }
}
