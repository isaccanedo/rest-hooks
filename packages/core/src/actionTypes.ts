export const FETCH_TYPE = 'rest-hooks/fetch' as const;
export const RECEIVE_TYPE = 'rest-hooks/receive' as const;
export const OPTIMISTIC_TYPE = 'rest-hooks/optimistic' as const;
export const RESET_TYPE = 'rest-hooks/reset' as const;
export const SUBSCRIBE_TYPE = 'rest-hooks/subscribe' as const;
export const UNSUBSCRIBE_TYPE = 'rest-hook/unsubscribe' as const;
export const INVALIDATE_TYPE = 'rest-hooks/invalidate' as const;
export const GC_TYPE = 'rest-hooks/gc' as const;
