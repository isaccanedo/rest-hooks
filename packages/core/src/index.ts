export * as __INTERNAL__ from './internal.js';
export type {
  NetworkError,
  UnknownError,
  ErrorTypes,
  Schema,
  EndpointInterface,
  EntityInterface,
  ResolveType,
} from '@rest-hooks/normalizr';
export { ExpiryStatus } from '@rest-hooks/normalizr';
export {
  default as NetworkManager,
  ResetError,
} from './state/NetworkManager.js';
export {
  default as createReducer,
  initialState,
} from './state/createReducer.js';
export { default as reducer } from './state/reducerInstance.js';
export { default as applyManager } from './state/applyManager.js';
export { useDenormalized } from './state/selectors/index.js';
export {
  useCache,
  useFetcher,
  useController,
  useFetchDispatcher,
  useRetrieve,
  useResource,
  useSubscription,
  useMeta,
  useError,
  CacheProvider,
  BackupBoundary,
  useInvalidator,
  useInvalidateDispatcher,
  useResetter,
  hasUsableData,
} from './react-integration/index.js';
export {
  useSuspense,
  useFetch,
  useDLE,
} from './react-integration/newhooks/index.js';
export {
  StateContext,
  DispatchContext,
  DenormalizeCacheContext,
  ControllerContext,
  StoreContext,
  type Store,
} from './react-integration/context.js';
export { default as Controller } from './controller/Controller.js';

export * from './controller/types.js';
export * from './state/actions/index.js';
export * as actionTypes from './actionTypes.js';
export { usePromisifiedDispatch } from '@rest-hooks/use-enhanced-reducer';
/* istanbul ignore next */
export * from './types.js';
export type {
  FetchShape,
  ReadShape,
  MutateShape,
  DeleteShape,
} from './endpoint/shapes.js';
export type {
  SetShapeParams,
  ParamsFromShape,
  BodyFromShape,
  ReturnFromShape,
} from './endpoint/types.js';
