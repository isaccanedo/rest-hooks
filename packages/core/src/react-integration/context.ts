import { createContext } from 'react';
import type { DenormalizeCache } from '@rest-hooks/normalizr';

import type { ActionTypes, State } from '../types.js';
import { initialState } from '../state/createReducer.js';
import Controller from '../controller/Controller.js';

export const StateContext = createContext(initialState);

const dispatch = (value: ActionTypes) => {
  /* istanbul ignore next */
  if (process.env.NODE_ENV !== 'production') {
    console.error(
      'It appears you are trying to use Rest Hooks without a provider.\nFollow instructions: https://resthooks.io/docs/getting-started/installation#add-provider-at-top-level-component',
    );
    /* istanbul ignore next */
    if (process.env.NODE_ENV === 'test') {
      console.error(
        'If you are trying to test: https://resthooks.io/docs/guides/unit-testing-hooks',
      );
    }
  }
  return Promise.resolve();
};
export const DispatchContext = createContext(dispatch);

// this is not needed anymore, but keeping around for backcompatibility
export const DenormalizeCacheContext = createContext<DenormalizeCache>({
  entities: {},
  results: {},
});

export const ControllerContext = createContext<Controller>(
  new Controller({
    dispatch,
    globalCache: {
      entities: {},
      results: {},
    },
  }),
);

export interface Store<S> {
  subscribe(listener: () => void): () => void;
  dispatch: React.Dispatch<ActionTypes>;
  getState(): S;
  uninitialized?: boolean;
}
/* istanbul ignore next */
export const StoreContext = createContext<Store<State<unknown>>>({
  subscribe: listener => () => {},
  dispatch: () => {},
  getState: () => initialState,
  uninitialized: true,
});
