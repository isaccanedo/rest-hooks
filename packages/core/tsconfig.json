{
  "extends": "../../tsconfig-base",
  "compilerOptions": {
    "outDir": "lib",
    "rootDir": "src"
  },
  "include": ["src"],
  "references": [
    { "path": "../../__tests__" },
    { "path": "../normalizr" },
    { "path": "../use-enhanced-reducer" }
  ]
}
