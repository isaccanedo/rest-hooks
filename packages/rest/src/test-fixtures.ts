export const payload = {
  id: 5,
  title: 'hi ho',
  content: 'whatever',
  tags: ['a', 'best', 'react'],
};

export const createPayload = {
  id: 1,
  title: 'hi ho',
  content: 'whatever',
  tags: ['a', 'best', 'react'],
};

export const articlesPages = {
  prevPage: '23asdl',
  nextPage: 's3f3',
  results: [
    {
      id: 23,
      title: 'the first draft',
      content: 'the best things in life com efree',
      tags: ['one', 'two'],
    },
    {
      id: 44,
      title: 'the second book',
      content: 'the best things in life com efree',
      tags: ['hbh', 'wew'],
    },
    {
      id: 2,
      title: 'the third novel',
      content: 'the best things in life com efree',
      tags: ['free', 'honey'],
    },
    {
      id: 643,
      title: 'a long time ago',
      content: 'the best things in life com efree',
    },
  ],
};

export const users = [
  {
    id: 23,
    username: 'bob',
    email: 'bob@bob.com',
    isAdmin: false,
  },
  {
    id: 7342,
    username: 'lindsey',
    email: 'lindsey@bob.com',
    isAdmin: true,
  },
];

export const nested = [
  {
    id: 5,
    title: 'hi ho',
    content: 'whatever',
    tags: ['a', 'best', 'react'],
    author: {
      id: 23,
      username: 'bob',
    },
  },
  {
    id: 3,
    title: 'the next time',
    content: 'whatever',
    author: {
      id: 23,
      username: 'charles',
      email: 'bob@bob.com',
    },
  },
];

export const moreNested = [
  {
    id: 7,
    title: 'article 7',
    content: 'whatever',
    tags: ['blah'],
    author: {
      id: 23,
      username: 'bob',
    },
  },
  {
    id: 8,
    title: 'article 8',
    content: 'whatever',
    author: {
      id: 27,
      username: 'zac',
      email: 'zac@bob.com',
    },
  },
];

export const valuesFixture = {
  first: {
    id: 1,
    title: 'first thing',
    content: 'blah',
    tags: [],
  },
  second: {
    id: 2,
    name: 'second thing',
    content: 'blah',
    tags: [],
  },
};

export const paginatedFirstPage = {
  nextPage: '2',
  data: { results: nested },
};

export const paginatedSecondPage = {
  nextPage: null,
  data: { results: moreNested },
};
