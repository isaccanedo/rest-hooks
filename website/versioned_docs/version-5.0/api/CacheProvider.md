---
title: "<CacheProvider />"
---

```typescript
interface ProviderProps {
  children: ReactNode;
  managers: Manager[];
  initialState: State<unknown>;
}
```

Manages state, providing all context needed to use the hooks. Should be placed as high as possible
in application tree as any usage of the hooks is only possible for components below the provider
in the React tree.

`index.tsx`

```tsx
import { CacheProvider } from 'rest-hooks';
import ReactDOM from 'react-dom';

ReactDOM.render(
  <CacheProvider>
    <App />
  </CacheProvider>,
  document.body
);
```

## initialState: State\<unknown\>

```typescript
type State<T> = Readonly<{
  entities: Readonly<{ [fetchKey: string]: { [pk: string]: T } | undefined }>;
  results: Readonly<{ [url: string]: unknown | PK[] | PK | undefined }>;
  meta: Readonly<{
    [url: string]: { date: number; error?: Error; expiresAt: number };
  }>;
}>;
```

Instead of starting with an empty cache, you can provide your own initial state. This can
be useful for testing, or rehydrating the cache state when using server side rendering.

## managers: Manager[]

Default:

```typescript
[new NetworkManager(), new SubscriptionManager(PollingSubscription)]
```

List of [Manager](./Manager)s use. This is the main extensibility point of the provider.
