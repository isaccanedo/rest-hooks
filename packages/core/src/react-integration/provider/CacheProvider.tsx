import React, { useMemo } from 'react';
import { useRef } from 'react';

import { initialState as defaultState } from '../../state/createReducer.js';
import NetworkManager from '../../state/NetworkManager.js';
import { State, Manager } from '../../types.js';
import { DenormalizeCacheContext, ControllerContext } from '../context.js';
import applyManager from '../../state/applyManager.js';
import CacheStore from './CacheStore.js';
import Controller from '../../controller/Controller.js';

/* istanbul ignore next  */
const SSR = typeof window === 'undefined';

interface ProviderProps {
  children: React.ReactNode;
  managers: Manager[];
  initialState: State<unknown>;
  Controller: typeof Controller;
}

/**
 * Controller managing state of the cache and coordinating network requests.
 * @see https://resthooks.io/docs/api/CacheProvider
 */
export default function CacheProvider({
  children,
  managers,
  initialState,
  Controller,
}: ProviderProps) {
  /* istanbul ignore else */
  if (process.env.NODE_ENV !== 'production' && SSR) {
    console.warn(
      `CacheProvider does not update while doing SSR.
Try using https://resthooks.io/docs/api/ExternalCacheProvider for server entry points.`,
    );
  }
  // contents of this component expected to be relatively stable

  const controllerRef: React.MutableRefObject<Controller> = useRef<any>();
  if (!controllerRef.current) controllerRef.current = new Controller();

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const memodManagers = useMemo(() => managers, managers);

  const middlewares = useMemo(
    () => applyManager(managers, controllerRef.current),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    managers,
  );

  return (
    <ControllerContext.Provider value={controllerRef.current}>
      <DenormalizeCacheContext.Provider
        value={controllerRef.current.globalCache}
      >
        <CacheStore
          managers={memodManagers}
          middlewares={middlewares}
          initialState={initialState}
          controller={controllerRef.current}
        >
          {children}
        </CacheStore>
      </DenormalizeCacheContext.Provider>
    </ControllerContext.Provider>
  );
}
CacheProvider.defaultProps = {
  managers: [new NetworkManager()] as Manager[],
  initialState: defaultState as State<unknown>,
  Controller,
};
