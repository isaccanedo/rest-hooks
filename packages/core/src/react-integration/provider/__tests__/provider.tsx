// eslint-env jest
import { CoolerArticle, CoolerArticleResource } from '__tests__/new';
import React, { useContext, Suspense } from 'react';
import { act, render } from '@testing-library/react';
import { NetworkManager, useSuspense } from '@rest-hooks/core';
import nock from 'nock';

import { RECEIVE_TYPE } from '../../../actionTypes';
import { DispatchContext, StateContext } from '../../context';
import CacheProvider from '../CacheProvider';
import { payload } from '../../test-fixtures';

describe('<CacheProvider />', () => {
  let warnspy: jest.SpyInstance;
  beforeEach(() => {
    warnspy = jest.spyOn(global.console, 'warn');
  });
  afterEach(() => {
    warnspy.mockRestore();
  });

  beforeAll(() => {
    nock(/.*/)
      .persist()
      .defaultReplyHeaders({
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Headers': 'Access-Token',
        'Content-Type': 'application/json',
      })
      .options(/.*/)
      .reply(200)
      .get(`/article-cooler/${payload.id}`)
      .reply(200, payload);
  });

  afterAll(() => {
    nock.cleanAll();
  });

  it('should warn users about missing Suspense', () => {
    const Component = () => {
      const article = useSuspense(CoolerArticleResource.get, { id: 5 });
      return <div>{article.title}</div>;
    };
    const tree = (
      <CacheProvider>
        <Component />
      </CacheProvider>
    );
    const { getByText } = render(tree);
    const msg = getByText('Uncaught Suspense.');
    expect(msg).toBeDefined();
    expect(warnspy).toHaveBeenCalled();
    expect(warnspy.mock.calls).toMatchSnapshot();
  });

  it('should not warn if Suspense is provided', () => {
    const Component = () => {
      const article = useSuspense(CoolerArticleResource.get, { id: 5 });
      return <div>{article.title}</div>;
    };
    const tree = (
      <CacheProvider>
        <Suspense fallback="loading">
          <Component />
        </Suspense>
      </CacheProvider>
    );
    const { getByText } = render(tree);
    const msg = getByText('loading');
    expect(msg).toBeDefined();
    expect(warnspy).not.toHaveBeenCalled();
  });

  it('should not change dispatch function on re-render', () => {
    let dispatch;
    let count = 0;
    function DispatchTester() {
      dispatch = useContext(DispatchContext);
      count++;
      return null;
    }
    const chil = <DispatchTester />;
    const tree = <CacheProvider>{chil}</CacheProvider>;
    const { rerender } = render(tree);
    expect(dispatch).toBeDefined();
    let curDisp = dispatch;
    rerender(tree);
    expect(curDisp).toBe(dispatch);
    expect(count).toBe(1);
    rerender(<CacheProvider>{chil}</CacheProvider>);
    expect(curDisp).toBe(dispatch);
    expect(count).toBe(1);
    const managers: any[] = [new NetworkManager()];
    rerender(<CacheProvider managers={managers}>{chil}</CacheProvider>);
    expect(count).toBe(2);
    curDisp = dispatch;
    rerender(<CacheProvider managers={managers}>{chil}</CacheProvider>);
    expect(curDisp).toBe(dispatch);
    expect(count).toBe(2);
    rerender(
      <DispatchContext.Provider value={() => Promise.resolve()}>
        {chil}
      </DispatchContext.Provider>,
    );
    expect(curDisp).not.toBe(dispatch);
    expect(count).toBe(3);
  });
  it('should change state', () => {
    let dispatch: any, state;
    let count = 0;
    function ContextTester() {
      dispatch = useContext(DispatchContext);
      state = useContext(StateContext);
      count++;
      return null;
    }
    const chil = <ContextTester />;
    const tree = <CacheProvider>{chil}</CacheProvider>;
    render(tree);
    expect(dispatch).toBeDefined();
    expect(state).toBeDefined();
    const action = {
      type: RECEIVE_TYPE,
      payload: { id: 5, title: 'hi', content: 'more things here' },
      meta: {
        schema: CoolerArticle,
        key: CoolerArticleResource.get.key({ id: 5 }),
        mutate: false,
        date: 50,
        expiresAt: 55,
      },
    };
    act(() => {
      dispatch(action);
    });
    expect(count).toBe(2);
    expect(state).toMatchSnapshot();
  });
});
