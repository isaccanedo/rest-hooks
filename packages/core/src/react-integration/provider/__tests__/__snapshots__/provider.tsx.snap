// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CacheProvider /> should change state 1`] = `
{
  "entities": {
    "CoolerArticle": {
      "5": {
        "content": "more things here",
        "id": 5,
        "title": "hi",
      },
    },
  },
  "entityMeta": {
    "CoolerArticle": {
      "5": {
        "date": 50,
        "expiresAt": 55,
        "fetchedAt": 50,
      },
    },
  },
  "indexes": {},
  "lastReset": 0,
  "meta": {
    "GET http://test.com/article-cooler/5": {
      "date": 50,
      "expiresAt": 55,
      "prevExpiresAt": undefined,
    },
  },
  "optimistic": [],
  "results": {
    "GET http://test.com/article-cooler/5": "5",
  },
}
`;

exports[`<CacheProvider /> should warn users about missing Suspense 1`] = `
[
  [
    "Uncaught suspense.
Make sure to add your own Suspense boundaries: https://resthooks.io/docs/getting-started/data-dependency#async-fallbacks",
  ],
]
`;
